mongo/client.go

```diff
@@ -136,13 +131,83 @@ func NewClient(opts ...*options.ClientOptions) (*Client, error) {
 	}
 	client := &Client{id: id}
 
-	err = client.configure(clientOpt)
+	// ClusterClock
+	client.clock = new(session.ClusterClock)
+
+	// LocalThreshold
+	client.localThreshold = defaultLocalThreshold
+	if clientOpt.LocalThreshold != nil {
+		client.localThreshold = *clientOpt.LocalThreshold
+	}
+	// Monitor
+	if clientOpt.Monitor != nil {
+		client.monitor = clientOpt.Monitor
+	}
+	// ServerMonitor
+	if clientOpt.ServerMonitor != nil {
+		client.serverMonitor = clientOpt.ServerMonitor
+	}
+	// ReadConcern
+	client.readConcern = readconcern.New()
+	if clientOpt.ReadConcern != nil {
+		client.readConcern = clientOpt.ReadConcern
+	}
+	// ReadPreference
+	client.readPreference = readpref.Primary()
+	if clientOpt.ReadPreference != nil {
+		client.readPreference = clientOpt.ReadPreference
+	}
+	// Registry
+	client.registry = bson.DefaultRegistry
+	if clientOpt.Registry != nil {
+		client.registry = clientOpt.Registry
+	}
+	// RetryWrites
+	client.retryWrites = true // retry writes on by default
+	if clientOpt.RetryWrites != nil {
+		client.retryWrites = *clientOpt.RetryWrites
+	}
+	client.retryReads = true
+	if clientOpt.RetryReads != nil {
+		client.retryReads = *clientOpt.RetryReads
+	}
+	// Timeout
+	client.timeout = clientOpt.Timeout
+	// WriteConcern
+	if clientOpt.WriteConcern != nil {
+		client.writeConcern = clientOpt.WriteConcern
+	}
+	// AutoEncryptionOptions
+	if clientOpt.AutoEncryptionOptions != nil {
+		if err := client.configureAutoEncryption(clientOpt); err != nil {
+			return nil, err
+		}
+	} else {
+		client.cryptFLE = clientOpt.Crypt
+	}
+
+	// Deployment
+	if clientOpt.Deployment != nil {
+		client.deployment = clientOpt.Deployment
+	}
+
+	// Set default options
+	if clientOpt.MaxPoolSize == nil {
+		clientOpt.SetMaxPoolSize(defaultMaxPoolSize)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	cfg, err := topology.NewConfig(clientOpt, client.clock)
```

Comment:
Does calling `NewConfig` here instead of just `configure` satisfy GODRIVER-1714 (and, if so can we make sure the final commit message contains GODRIVER-1714)? Seems like we're now doing the same validation for creating a client from a URI as we do for creating a client from `ClientOption`s.

Reply:
GODRIVER-1714 is not resolved by this. `topology.NewConfig` deprecates the `tolopology.WithConnString` while merging most of the logic of configure into the topology package. However, a user could build a URI without using the connstring package and then pass it to the client options with `ClientOptions.ApplyURI`, this would bypass these validations. GODRIVER-1714 would require merging the validations from `connstring.ParseAndValidate` into `ClientOptions.Validate`.